src/e2ee.rs

```diff
@@ -290,36 +291,60 @@ async fn decrypt_if_autocrypt_message(
         encrypted_data_part,
         private_keyring,
         public_keyring_for_validate,
-        ret_valid_signatures,
     )
     .await
 }
 
+/// Validates signatures of Multipart/Signed message part, as defined in RFC 1847.
+///
+/// Returns `None` if the part is not a Multipart/Signed part, otherwise retruns the set of key
+/// fingerprints for which there is a valid signature.
+async fn validate_detached_signature(
+    mail: &ParsedMail<'_>,
+    public_keyring_for_validate: &Keyring<SignedPublicKey>,
+) -> Result<Option<(Vec<u8>, HashSet<Fingerprint>)>> {
+    if mail.ctype.mimetype != "multipart/signed" {
+        return Ok(None);
+    }
+
+    if let [first_part, second_part] = &mail.subparts[..] {
+        // First part is the content, second part is the signature.
+        let content = first_part.raw_bytes;
+        let signature = second_part.get_body_raw()?;
+        let ret_valid_signatures =
+            pgp::pk_validate(content, &signature, public_keyring_for_validate).await?;
+
+        Ok(Some((content.to_vec(), ret_valid_signatures)))
+    } else {
+        Ok(None)
+    }
+}
+
 /// Returns Ok(None) if nothing encrypted was found.
 async fn decrypt_part(
     mail: &ParsedMail<'_>,
     private_keyring: Keyring<SignedSecretKey>,
     public_keyring_for_validate: Keyring<SignedPublicKey>,
-    ret_valid_signatures: &mut HashSet<Fingerprint>,
-) -> Result<Option<Vec<u8>>> {
+) -> Result<Option<(Vec<u8>, HashSet<Fingerprint>)>> {
     let data = mail.get_body_raw()?;
 
     if has_decrypted_pgp_armor(&data) {
-        // we should only have one decryption happening
-        ensure!(ret_valid_signatures.is_empty(), "corrupt signatures");
-
-        let plain = pgp::pk_decrypt(
-            data,
-            private_keyring,
-            public_keyring_for_validate,
-            Some(ret_valid_signatures),
-        )
-        .await?;
-
-        // If the message was wrongly or not signed, still return the plain text.
-        // The caller has to check the signatures then.
+        let (plain, ret_valid_signatures) =
+            pgp::pk_decrypt(data, private_keyring, &public_keyring_for_validate).await?;
+
+        // Check for detached signatures.
+        // If decrypted part is a multipart/signed, then there is a detached signature.
+        let decrypted_part = mailparse::parse_mail(&plain)?;
```

Comment:
the additional call to `mailparse::parse_mail()` is executed on each message, right? even if it has a signature following Autocrypt standards?  

wondering how expensive that is and if if make sense to return the results to the caller so that we do not need two calls to `mailparse::parse_mail()` (that may save the call at https://github.com/deltachat/deltachat-core-rust/blob/master/src/mimeparser.rs#L214)

but i did only a rough check, not sure if the input data for `mailparse::parse_mail()` are exactly the same and how fast or slow `mailparse::parse_mail()` actually is.

Reply:
I have tried to do it, but it is hard to return both the owned decrypted plaintext and parsed mail that references this decrypted plaintext from `decrypt_part`. For this we need an owned `ParsedMail` variant that stores both the decrypted data and the parsed MIME tree pointing into it. Currently `ParsedMail` contains `&[u8]` inside of it, and we need the same structure but containing `Vec<u8>`. It is possible to build one by making a structure `OwnedParsedMail` with `Vec<u8>` and `ParsedMail` pointing to it inside with [ourobors](https://lib.rs/crates/ouroboros) crate, and even implement `Cow` and `.to_owned()` for `ParsedMail` then, so `ParsedMail::Owned` is an `OwnedParsedMail`, but this should really be coded into `mailparse` crate then.

Reply:
I opened an issue in the `mailparse` repo with these notes, but think it's nothing urgent, just nice-to-have: https://github.com/staktrace/mailparse/issues/97

Reply:
thanks for taking care! might also be that this double-decoding is few overhead for most messages that are only a few kb large. but, as you said, nice to have :)